implementations/rust/ockam/ockam_core/src/api.rs

```diff
@@ -587,31 +576,60 @@ impl Request<()> {
     }
 }
 
-impl<T: Encode<()>> Request<T> {
-    pub fn encode<W>(&self, buf: W) -> Result<(), encode::Error<W::Error>>
+impl<T: Encodable> Request<T> {
+    fn encode_request<W>(self, buf: W) -> Result<(), encode::Error<W::Error>>
     where
         W: Write,
     {
         let mut e = Encoder::new(buf);
         e.encode(&self.header)?;
-        if let Some(b) = &self.body {
-            e.encode(b)?;
+        if let Some(b) = self.body {
+            e.writer_mut()
+                .write_all(&<T as Encodable>::encode(b).map_err(encode::Error::message)?)
+                .map_err(|_| encode::Error::message("encoding error"))?;
         }
         Ok(())
     }
 
-    pub fn to_vec(&self) -> Result<Vec<u8>, encode::Error<<Vec<u8> as Write>::Error>> {
+    fn into_vec(self) -> Result<Vec<u8>, encode::Error<<Vec<u8> as Write>::Error>> {
         let mut buf = Vec::new();
-        self.encode(&mut buf)?;
-
+        self.encode_request(&mut buf)?;
         Ok(buf)
     }
 }
 
-#[derive(Debug)]
+impl<T: Encodable> Encodable for Request<T> {
+    /// This double serialization of requests with both
+    /// cbor and then serde_bare is not necessary and could be removed
+    fn encode(self) -> Result<Encoded> {
+        serialize(self.into_vec()?)
+    }
+}
+
+impl<T: Decodable> Decodable for Request<T> {
+    fn decode(e: &[u8]) -> Result<Self> {
+        let deserialized = deserialize::<Vec<u8>>(e)?;
+        let mut dec = Decoder::new(&deserialized);
+        let header: RequestHeader = dec.decode()?;
+        if header.has_body() {
+            let body = dec.input().get(dec.position()..deserialized.len()).unwrap();
```

Comment:
Since the request's body is an option, wouldn't it be better to default to None if we fail to get the body, instead of unwrapping?

Reply:
I should return a better error here, but this should still be an error if we can't get the rest of the input.

Reply:
Done.